R/plotNorm.R

```diff
@@ -76,25 +80,141 @@ plotNorm <- function(DGEdata,
-        resultPlot <- ggplot(tall, aes(x = Log2CPM, color = SampleID)) +
-            geom_density() +
-            facet_grid(~Normalization) +
-            ggtitle(stringr::str_c("Log2CPM before/after", normalize, "normalization", sep = " "))  +
-            theme_gray() +
-            theme(legend.position = "none")
+    if (tolower(plotCategory) == "density") {
+        if (plotType == "canvasXpress") {
+            resultPlot <- build_cx_density_plot(tall, title)
+        } else {
+            resultPlot <- build_gg_density_plot(tall, title)
+        }
+    } else {
+        if (plotType == "canvasXpress") {
+            resultPlot <- build_cx_box_plot(tall, title)
+        } else {
+            resultPlot <- build_gg_box_plot(tall, title)
+        }
     }
 
-    if (tolower(plotType == "box")) {
-        resultPlot <- ggplot(tall, aes(x = SampleID, y = Log2CPM, color = SampleID)) +
-            geom_boxplot(alpha = 0.5) +
-            facet_grid(~Normalization) +
-            ggtitle(stringr::str_c("Log2CPM before/after", normalize, "normalization", sep = " "))  +
-            theme_gray() +
-            theme(axis.text.x = element_blank(),
-                  legend.position = "none")
+    return(resultPlot)
+}
+
+build_cx_density_data <- function(data) {
+    cx.data <- data.frame()
+    for (sample in levels(as.factor(data$SampleID))) {
+        if (nrow(cx.data) == 0) {
+            cx.data <- select_sample(data, sample, TRUE)
+        } else {
+            cx.data <-
+                cbind(cx.data, select_sample(data, sample))
+        }
+
     }
+    return(cx.data)
+}
+
+select_sample <- function(data, sampleID, select_all = FALSE) {
+    if (select_all) {
+        sample <-
+            subset(data, SampleID == sampleID,
+                   select = c("GeneID", "Normalization", "Log2CPM"))
+        colnames(sample)[3] = sampleID
+    } else {
+        sample <-
+            subset(data, SampleID == sampleID, select = c("Log2CPM"))
+        colnames(sample)[1] = sampleID
+    }
+
+    return(sample)
+}
+
+build_cx_density_plot <- function(data, title) {
+    cx.data <- build_cx_density_data(data)
+    xlab <- "Log2CPM"
+    ylab <- "density"
+    events <- htmlwidgets::JS(
+        "{ 'mousemove' : function(o, e, t) {
+                           if (o != null && o != false) {
+                             if (o.objectType == 'Density' &&
+                                o.display != null &&
+                                o.display.indexOf('-') > -1) {
+                                sampleID = o.display.split('-')[2];
+                                t.showInfoSpan(e,
+                                '<b>Sample ID</b>: ' + sampleID );
+                             } else {
+                               t.showInfoSpan(e, o.display);
+                             };
+                          }; }}")
+    resultPlot <- canvasXpress::canvasXpress(data                    = cx.data[,-c(1, 2)],
```

Comment:
Index-based subsetting often creates problems. I would suggest refactor code to subset based on colnames .